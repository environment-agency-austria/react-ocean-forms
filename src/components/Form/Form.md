### Examples

#### Simple form

Showcase of a simple form

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // Reset callback, can be useful in some cases for cleanup
  const handleReset = () => {
    console.log('onReset');
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} onReset={handleReset}>
      <Input name="myInput" label="Example input" />
      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
    </Form>
  );
}

<Example />;
```

#### Default values

Provide the form with default values that should be displayed on load. Note how the default value won't update the Field, if the user changed the input. However, on form reset the default value will be used again.

```jsx
import React, { useState } from 'react';
import { Form, Input } from 'react-ocean-forms';

function Example() {
  const [defaultValues, setDefaultValues] = useState({ myInput: 'default value' });

  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  const randomizeDefaultValues = () => {
    setDefaultValues({
      myInput: `default ${Math.floor(Math.random() * 100)}`,
    });
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} defaultValues={defaultValues}>
      <Input name="myInput" label="Example input" />

      <p className="mt-4">Current default value: {defaultValues.myInput}</p>

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
      <button type="button" onClick={randomizeDefaultValues} className="ml-1">
        Update default values
      </button>
    </Form>
  );
}

<Example />;
```

#### Values

Override the values of the form fields. Changing those values will override the Field value, even if the user changed it.

```jsx
import React, { useState } from 'react';
import { Form, Input } from 'react-ocean-forms';

function Example() {
  const [values, setValues] = useState({ myInput: 'demo value' });

  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  const randomizeValues = () => {
    setValues({
      myInput: `demo ${Math.floor(Math.random() * 100)}`,
    });
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} values={values}>
      <Input name="myInput" label="Example input" />

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
      <button type="button" onClick={randomizeValues} className="ml-1">
        Update values
      </button>
    </Form>
  );
}

<Example />;
```

#### Disabled form

Setting the disabled prop will disable all form fields.

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} disabled>
      <Input name="myInput" label="Example input 1" />
      <Input name="myInput2" label="Example input 2" />
      <button type="submit">Submit</button>
    </Form>
  );
}

<Example />;
```

#### Plaintext form

Setting the plaintext prop will show all form fields in a text-only mode.

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example({ logMessage }) {
  const defaultValues = {
    myInput: 'default value 1',
    myInput2: 'default value 2',
  };

  return (
    <Form className="demo" defaultValues={defaultValues} plaintext>
      <Input name="myInput" label="Example input 1" />
      <Input name="myInput2" label="Example input 2" />
    </Form>
  );
}

<Example />;
```

#### Form-wide validation

Use a form-wide validation function before submit. Notice that the onSubmit callback is not invoked if you type 'bad' into the input.

```jsx
import { Form, Input, ValidationSummary } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = values => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // Reset callback, can be useful in some cases for cleanup
  const handleReset = () => {
    console.log('onReset');
  };

  // Form wide validation function
  const handleValidate = values => {
    if (values.demoInput === 'bad') {
      return {
        demoInput: 'Invalid input!',
      };
    }

    return null;
  };

  return (
    <Form
      className="demo"
      onSubmit={handleSubmit}
      onReset={handleReset}
      onValidate={handleValidate}
    >
      <ValidationSummary id="summary" />

      <Input name="demoInput" label="Example input" />

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
    </Form>
  );
}

<Example />;
```
