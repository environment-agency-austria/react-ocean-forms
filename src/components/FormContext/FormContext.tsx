/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';
import { IFormContext } from './FormContext.types';

/**
 * Context for the communication between the form
 * and its fields.
 */
export const FormContext = React.createContext<IFormContext | undefined>(undefined);
