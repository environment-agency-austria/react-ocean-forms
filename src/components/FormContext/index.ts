export * from './FormContext';
export * from './FormContext.types';
export * from './FormContext.hooks';
