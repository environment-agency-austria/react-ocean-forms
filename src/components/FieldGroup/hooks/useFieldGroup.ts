/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module FieldGroup
 */
import { useCallback, useMemo } from 'react';

import { useFormContext, useValidation, IValidationArgs, IBasicValidationState } from '../../../hooks';
import { useFullName, useFieldRegistration } from '../../../hooks/internal';
import { IFormContext, IFieldValues } from '../../FormContext';

import { getGroupValue } from './useFieldGroup.utils';
import { IUseFieldGroupArgs, IUseFieldGroupResult } from './useFieldGroup.types';

/**
 * @hidden
 */
export function useFieldGroup<TFieldValue = unknown>(props: IUseFieldGroupArgs<TFieldValue>): IUseFieldGroupResult {
  const formContext = useFormContext();

  const {
    name,
    label,
    defaultValues,
    values,
    disabled = formContext.disabled,
    plaintext = formContext.plaintext,
    asyncValidateOnChange = formContext.asyncValidateOnChange,
  } = props;

  const fullName = useFullName(name);
  const { validationState, validate, resetValidation, updateValidationState } = useValidation(props);

  /**
   * Triggers the validation of the group
   */
  const validateGroup = useCallback(
    (args?: Partial<IValidationArgs>): Promise<IBasicValidationState> => {
      const groupValue = getGroupValue<TFieldValue>(formContext, fullName);
      return validate(groupValue, args);
    },
    [formContext, fullName, validate],
  );

  // Register the group in the formContext, so the group
  // validation can be called on form submit.
  const registerFieldState = useMemo(() => ({
    label,
    isGroup: true,
    updateValidation: updateValidationState,
    validate: validateGroup,
    reset: resetValidation,
    getValue: () => ({}),
  }), [label, resetValidation, updateValidationState, validateGroup]);
  useFieldRegistration(
    fullName,
    registerFieldState,
  );

  /**
   * Listens to child field events, triggers validation if
   * needed and passes them to the higher context
   * @param name Field name
   * @param event Event name
   * @param args Event args
   */
  const notifyFieldEvent = useCallback(
    (name: string, event: string, args?: unknown): void => {
      formContext.notifyFieldEvent(name, event, args);

      if (event !== 'change' && event !== 'blur') { return; }

      if (event === 'change') {
        const localName = name.substring(fullName.length + 1);

        const currentGroupValue = getGroupValue<TFieldValue>(formContext, fullName);
        const intermediateGroupValue = {
          ...(currentGroupValue === undefined ? { } : currentGroupValue),
          ...{
            // Override the value of the event sender, because
            // the Field didn't update its state yet, making the
            // Form.getValues() returning an old Field value.
            [localName]: args,
          },
        };

        void validate(
          intermediateGroupValue as TFieldValue,
          { checkAsync: asyncValidateOnChange },
        );
      } else if (!asyncValidateOnChange) {
        void validate(getGroupValue(formContext, fullName));
      }
    },
    [asyncValidateOnChange, formContext, fullName, validate],
  );

  const subContext: IFormContext = useMemo(() => ({
    ...formContext,
    fieldPrefix: fullName,
    notifyFieldEvent,
    disabled,
    plaintext,
    defaultValues: (defaultValues === undefined ? formContext.defaultValues : { ...formContext.defaultValues, ...{ [fullName]: defaultValues }}) as Partial<IFieldValues>,
    values: (values === undefined ? formContext.values : { ...formContext.values, ...{ [fullName]: values }}) as Partial<IFieldValues>,
  }), [defaultValues, disabled, formContext, fullName, notifyFieldEvent, plaintext, values]);

  const groupState = useMemo(
    () => ({
      fullName,
      isValidating: validationState.isValidating,
      isRequired: validationState.isRequired,
      valid: validationState.valid,
      error: validationState.error,
    }),
    [
      fullName,
      validationState,
    ],
  );

  return {
    groupFormContext: subContext,
    renderParams: groupState,
  };
}
