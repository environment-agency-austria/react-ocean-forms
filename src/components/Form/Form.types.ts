/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { TSTringFormatter } from '../../utils';
import { TFieldError } from '../../validators';
import { IFieldValues } from '../FormContext';

/**
 * Props for the Form component
 */
export interface IFormProps<TFieldValues = IFieldValues, TSubmitArgs = unknown> {
  /**
   * Contains the default values of the form. Those values will be
   * put into the according fields when the form initializes.
   */
  defaultValues: Partial<TFieldValues>;
  /**
   * Contains the values of the form. Changing this property will
   * update all Field values, overwriting their default values but also
   * any value the user put in.
   */
  values?: Partial<TFieldValues>;
  /**
   * If set to true the form will trigger asynchronous validation on
   * Fields whenever they change (e.g. on key press). Default behaviour
   * is that the fields will only async validate when they loose focus.
   * Can be overriden per field.
   */
  asyncValidateOnChange: boolean;
  /**
   * Configures the wait time before the async validators will be called.
   * Per default the form will call the async validators only 400ms after
   * the last value change. Can be overriden per field.
   */
  asyncValidationWait: number;
  /**
   * If set every text output will be put through this function.
   * Per default an internal implementation is used.
   */
  formatString: TSTringFormatter;
  /**
   * If set to true the form will disable all Fields and FormButtons.
   */
  disabled: boolean;
  /**
   * Sets the className of the html form.
   */
  className?: string;
  /**
   * If set to true, all the fields will display only text instead of an
   * input element. This is useful to re-use Fields in a check page.
   */
  plaintext: boolean;
  /**
   * If set to true, all fields will be reset on an successful form submit
   */
  resetOnSubmit?: boolean;
  /**
   * If set to true, the form will be forced into a busy state and thus disabling
   * any form buttons.
   */
  busy?: true;

  /**
   * Triggered when the form has been validated successfully and is ready to be submitted.
   * @param values Contains the form values. The name of the fields are
   * used as property names for the values object. FieldGroups result in a nested object.
   * @param submitArgs By default undefined. Can be set by FormButton or
   * any other manual way of calling the submit method of the form context.
   */
  onSubmit?(values: TFieldValues, submitArgs?: TSubmitArgs): Promise<void> | void;
  /**
   * Triggered after all field validations have been successfull. Provides the current
   * values end expects an error object with the field names as properties and the errors
   * inside those properties.
   * @param values Contains the form values. The name of the fields are used as property
   * names for the values object. FieldGroups result in a nested object.
   */
  onValidate?(values: TFieldValues): TFormValidationResult;
  /**
   * Triggered when the form has been resetted by the user.
   */
  onReset?(): void;
}

export type TFormValidationResult = {
  [prop: string]: TFieldError | TFormValidationResult;
} | null;
