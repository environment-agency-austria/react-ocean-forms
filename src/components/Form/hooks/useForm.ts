/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module Form
 */
import { useMemo, useState, useCallback } from 'react';

import {
  stringFormatter as defaultStringFormatter,
  noopFunction,
  parseValidationError,
  getDeepValue,
} from '../../../utils';
import { IBasicValidationState } from '../../../hooks';
import { IFormContext, IFieldValues } from '../../FormContext';
import { IFormProps } from '../Form.types';
import { useFieldEvents, useFieldStates, useIsUnmounted } from '../../../hooks/internal';

/**
 * @hidden
 */
export function useForm<TFieldValues extends Record<string, unknown> = IFieldValues>(
  props: IFormProps<TFieldValues>
): IFormContext<TFieldValues> {
  const [busyState, setBusyState] = useState(false);

  const {
    defaultValues = {},
    values,
    asyncValidationWait = 400,
    asyncValidateOnChange = false,
    formatString = defaultStringFormatter,
    disabled = false,
    plaintext = false,
    busy = busyState,
    resetOnSubmit,
    onReset = noopFunction,
    onSubmit = noopFunction,
    onValidate = noopFunction,
  } = props;

  const { registerListener, unregisterListener, notifyListeners } = useFieldEvents();
  const { getFieldState, registerField, unregisterField, forEachFieldState } = useFieldStates();
  const isUnmounted = useIsUnmounted();

  /**
   * Generates and returns an object that contains
   * all values from all the fields.
   * @returns Current values in form of { name: value, name2: value2, ... }
   */
  const getValues = useCallback((): TFieldValues => {
    const values = {} as TFieldValues;

    forEachFieldState((state, name) => {
      if (state.isGroup === true) {
        return;
      }

      const nameParts = name.split('.');
      let valueRef: Record<string, unknown> = values;

      nameParts.forEach((key, index) => {
        if (nameParts.length === 1 || index === nameParts.length - 1) {
          valueRef[key] = state.getValue();
        } else {
          if (valueRef[key] === undefined) {
            valueRef[key] = {};
          }
          valueRef = valueRef[key] as TFieldValues;
        }
      });
    });

    return values;
  }, [forEachFieldState]);

  /**
   * Sets the state of all fields back
   * to the default state.
   */
  const reset = useCallback((): void => {
    forEachFieldState(state => {
      state.reset();
    });

    onReset();
  }, [forEachFieldState, onReset]);

  /**
   * Triggers the form wide validation callback if given
   * @returns Validation state of the form
   */
  const triggerFormValidation = useCallback((): boolean => {
    const values = getValues();
    const result = onValidate(values);

    // If the callback returned null then the form is valid
    if (result === null || result === undefined) {
      return true;
    }

    // Otherwise parse the result object and update the
    // field states.
    let allFieldsValid = true;
    forEachFieldState((state, name) => {
      const fieldError = parseValidationError(getDeepValue(name, result));
      const isValid = fieldError === null || typeof fieldError !== 'object';

      if (isValid) {
        return;
      }

      state.updateValidation({
        valid: false,
        error: fieldError,
      });
      allFieldsValid = false;
    });

    return allFieldsValid;
  }, [forEachFieldState, getValues, onValidate]);

  /**
   * Submits the form - triggers any validations if needed
   * and raises the onFormSubmit prop callback if the
   * form is currently valid.
   * @param submitArgs Arguments that will be passed
   * to the onSubmit callback
   */
  const submit = useCallback(
    async (submitArgs?: unknown): Promise<void> => {
      setBusyState(true);

      // Iterate through all fields and validate them
      // if needed.
      const validations: Promise<IBasicValidationState>[] = [];
      forEachFieldState(field => {
        validations.push(
          field.validate({
            checkAsync: true,
            immediateAsync: true,
          })
        );
      });
      const validationStates = await Promise.all(validations);

      const notifyInvalid = (): void => {
        notifyListeners('_form', 'submit-invalid');
        setBusyState(false);
      };

      // Check if all fields are valid
      const allValid = validationStates.every(state => state.valid === true);
      if (allValid === false) {
        notifyInvalid();
        return;
      }

      // Call the form wide validation
      const formValid = triggerFormValidation();
      if (formValid === false) {
        notifyInvalid();
        return;
      }

      // Await the result from callOnSubmit
      const callOnSubmitResult = onSubmit(getValues(), submitArgs);

      // Make sure the state is cleaned up before
      const cleanup = (resetForm: boolean | undefined): void => {
        if (isUnmounted.current) return;

        setBusyState(false);
        if (resetForm) {
          reset();
        }
      };
      if (callOnSubmitResult instanceof Promise) {
        void callOnSubmitResult.then(() => {
          cleanup(resetOnSubmit);
        });
      } else {
        cleanup(resetOnSubmit);
      }
    },
    [
      forEachFieldState,
      getValues,
      isUnmounted,
      notifyListeners,
      onSubmit,
      reset,
      resetOnSubmit,
      triggerFormValidation,
    ]
  );

  const formContext: IFormContext<TFieldValues> = useMemo(
    () => ({
      fieldPrefix: null,
      defaultValues,
      values,
      asyncValidationWait,
      stringFormatter: formatString,
      disabled,
      plaintext,
      busy,
      asyncValidateOnChange,

      registerField,
      unregisterField,

      notifyFieldEvent: notifyListeners,
      registerListener,
      unregisterListener,

      getFieldState: getFieldState,
      getValues: getValues,

      reset,
      submit,
    }),
    [
      asyncValidateOnChange,
      asyncValidationWait,
      busy,
      defaultValues,
      disabled,
      formatString,
      getFieldState,
      getValues,
      notifyListeners,
      plaintext,
      registerField,
      registerListener,
      reset,
      submit,
      unregisterField,
      unregisterListener,
      values,
    ]
  );

  return formContext;
}
