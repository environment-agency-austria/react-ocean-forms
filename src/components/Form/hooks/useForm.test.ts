import { renderHook, act } from 'react-hooks-testing-library';

import { IFormProps } from '../Form.types';
import { IFormContext, IFieldValues } from '../../FormContext';

import { useForm } from './useForm';
import { useFieldEvents, IUseFieldEventsResult, IUseFieldStatesResult, useFieldStates, IFieldState } from '../../../hooks/internal';

jest.mock('../../../hooks/internal');

describe('useForm', () => {
  interface ISetupArgs {
    props?: IFormProps;
    fieldStatesOverride?: Partial<IUseFieldStatesResult>;
  }

  interface IMockResult {
    current: IFormContext;
  }

  interface ISetupResult {
    result: IMockResult;
    fieldEventResult: IUseFieldEventsResult;
    fieldStatesResult: IUseFieldStatesResult;
  }

  const setup = ({ props = { }, fieldStatesOverride }: Partial<ISetupArgs> = {}): ISetupResult => {
    const fieldEventResult: IUseFieldEventsResult = {
      notifyListeners: jest.fn(),
      registerListener: jest.fn(),
      unregisterListener: jest.fn(),
    };
    (useFieldEvents as jest.Mock).mockReturnValue(fieldEventResult);

    const fieldStatesResult: IUseFieldStatesResult = {
      registerField: jest.fn(),
      unregisterField: jest.fn(),
      getFieldState: jest.fn(),
      forEachFieldState: jest.fn(),
      ...fieldStatesOverride,
    };
    (useFieldStates as jest.Mock).mockReturnValue(fieldStatesResult);

    return {
      fieldEventResult,
      fieldStatesResult,
      ...renderHook(() => useForm(props)),
    };
  };

  const createMockFieldState = (label: string, isGroup?: boolean): IFieldState => ({
    label,
    isGroup,
    validate: jest.fn().mockResolvedValue({
      isValidating: false,
      valid: true,
      error: null,
    }),
    updateValidation: jest.fn(),
    reset: jest.fn(),
    getValue: jest.fn().mockReturnValue(label),
  });

  interface IMockField {
    name: string;
    state: IFieldState;
  }

  const createMockField = (name: string, label: string, isGroup?: boolean): IMockField => ({
    name,
    state: createMockFieldState(label, isGroup),
  });

  it('should create a valid form context', () => {
    const { result } = setup();
    expect(result.current).toMatchObject({
      fieldPrefix: null,
      registerField: expect.any(Function),
      unregisterField: expect.any(Function),
      notifyFieldEvent: expect.any(Function),
      registerListener: expect.any(Function),
      unregisterListener: expect.any(Function),
      getFieldState: expect.any(Function),
      getValues: expect.any(Function),
      busy: false,
      disabled: false,
      asyncValidateOnChange: false,
      asyncValidationWait: 400,
      defaultValues: {},
      values: undefined,
    });
  });

  describe('configuration', () => {
    const cases: any[] = [
      ['disabled', true, 'disabled'],
      ['formatString', jest.fn(), 'stringFormatter'],
      ['asyncValidateOnChange', true, 'asyncValidateOnChange'],
      ['asyncValidationWait', 800, 'asyncValidationWait'],
      ['defaultValues', { foo: 'bar' }, 'defaultValues'],
      ['values', { foo: 'bar' }, 'values'],
      ['plaintext', true, 'plaintext'],
    ];

    test.each(cases)('case %s', (prop, value, contextProp) => {
      const { result } = setup({
        props: { [prop]: value },
      });
      expect(result.current).toMatchObject({ [contextProp]: value });
    });
  });

  describe('formContext.getValues - form values', () => {
    const unitField = createMockField('unitField', 'Unit field');
    const unitGroup = createMockField('unitGroup', 'Unit group', true);
    const unitSubField = createMockField(`${unitGroup.name}.subField`, 'Sub field');
    const unitSubField2 = createMockField(`${unitGroup.name}.subField2`, 'Sub field 2');

    const mockFields = new Map<string, IFieldState>([
      [ unitField.name, unitField.state ],
      [ unitGroup.name, unitGroup.state ],
      [ unitSubField.name, unitSubField.state ],
      [ unitSubField2.name, unitSubField2.state ],
    ]);

    const forEachFieldState = jest.fn().mockImplementation((cb) => {
      mockFields.forEach(cb);
    });
    const { result } = setup({ fieldStatesOverride: { forEachFieldState }});

    let formValues: IFieldValues;
    it('should return the values without crashing', () => {
      expect(() => {
        formValues = result.current.getValues();
      }).not.toThrowError();
    });

    it('should call unitField.getValue', () => {
      expect(unitField.state.getValue).toHaveBeenCalled();
    });

    it('should call subField.getValue', () => {
      expect(unitSubField.state.getValue).toHaveBeenCalled();
    });

    it('should not call group.getValue', () => {
      // The value of fieldGroups is computed by its children
      expect(unitGroup.state.getValue).not.toHaveBeenCalled();
    });

    it('should return the correct form values', () => {
      const subFieldLocalName = unitSubField.name.substring(unitGroup.name.length + 1);
      const subFieldLocalName2 = unitSubField2.name.substring(unitGroup.name.length + 1);

      const expectedFormValues = {
        [unitField.name]: unitField.state.label,
        [unitGroup.name]: {
          [subFieldLocalName]: unitSubField.state.label,
          [subFieldLocalName2]: unitSubField2.state.label,
        },
      };

      expect(formValues).toMatchObject(expectedFormValues);
    });
  });

  describe('onSubmit handling', () => {
    const simulateSubmitEvent = async (context: IFormContext): Promise<void> => {
      return act(async () => {
        await context.submit();
      });
    };

    interface ISetupSubmitArgs extends ISetupArgs {
      customField: IMockField;
    }

    interface ISetupSubmitResult extends ISetupResult {
      expectedFormValues: IFieldValues;
      mockFields: Map<string, IFieldState>;
    }

    const unitFieldName = 'unitField';

    const setupSubmit = async ({ props, customField, }: Partial<ISetupSubmitArgs> = {}): Promise<ISetupSubmitResult> => {
      const unitField = createMockField(unitFieldName, 'Unit field');
      const unitGroup = createMockField('unitGroup', 'Unit group', true);
      const unitSubField = createMockField(`${unitGroup.name}.subField`, 'Sub field');

      const mockFields = new Map<string, IFieldState>([
        [ unitField.name, unitField.state ],
        [ unitGroup.name, unitGroup.state ],
        [ unitSubField.name, unitSubField.state ],
      ]);

      if (customField) {
        mockFields.set(customField.name, customField.state);
      }

      const forEachFieldState = jest.fn().mockImplementation((cb) => {
        mockFields.forEach(cb);
      });

      const result = setup({ props, fieldStatesOverride: { forEachFieldState } });

      await simulateSubmitEvent(result.result.current);

      const subFieldLocalName = unitSubField.name.substring(unitGroup.name.length + 1);
      const expectedFormValues = {
        [unitField.name]: unitField.state.label,
        [unitGroup.name]: {
          [subFieldLocalName]: unitSubField.state.label,
        },
      };

      return {
        ...result,
        expectedFormValues,
        mockFields,
      };
    };

    describe('all valid', () => {
      it('should call all the validation functions', async () => {
        const { mockFields } = await setupSubmit();

        mockFields.forEach(item => expect(item.validate).toHaveBeenLastCalledWith({
          checkAsync: true,
          immediateAsync: true,
        }));
      });

      it('should call the onValidate prop', async () => {
        const onValidateHandler = jest.fn().mockReturnValue(null);
        const { expectedFormValues } = await setupSubmit({ props: { onValidate: onValidateHandler }});

        expect(onValidateHandler).toHaveBeenCalledWith(expectedFormValues);
      });

      it('should call the onSubmit prop', async () => {
        const onSubmitHandler = jest.fn();
        const { expectedFormValues } = await setupSubmit({ props: { onSubmit: onSubmitHandler }});

        expect(onSubmitHandler).toHaveBeenCalledWith(expectedFormValues, undefined);
      });

      describe('onValidate handler that returns valid field states', () => {
        it('should call the onSubmit prop', async () => {
          const onValidateHandler = jest.fn().mockReturnValue({
            unitField: undefined,
          });
          const onSubmitHandler = jest.fn();
          const { expectedFormValues } = await setupSubmit({ props: { onValidate: onValidateHandler, onSubmit: onSubmitHandler }});

          expect(onSubmitHandler).toHaveBeenCalledWith(expectedFormValues, undefined);
        });
      });
    });

    describe('invalid through form validator', () => {
      const createInvalidValidator = (): jest.Mock => jest.fn().mockReturnValue({
        [unitFieldName]: 'error',
      });

      it('should update the validation state of the field', async () => {
        const { mockFields } = await setupSubmit({ props: { onValidate: createInvalidValidator() }});

        mockFields.forEach((item, name) => {
          if (name !== unitFieldName) { return; }
          expect(item.updateValidation).toHaveBeenCalledWith({
            valid: false,
            error: {
              message_id: 'error',
              params: { },
            },
          });
        });
      });

      it('should trigger a submit-invalid event', async () => {
        const { fieldEventResult } = await setupSubmit({ props: { onValidate: createInvalidValidator() } });
        expect(fieldEventResult.notifyListeners).toHaveBeenCalledWith('_form', 'submit-invalid');
      });

      it('should not call the onSubmit prop', async () => {
        const onSubmitHandler = jest.fn();
        await setupSubmit({ props: {
          onValidate: createInvalidValidator(),
          onSubmit: onSubmitHandler,
        }});
        expect(onSubmitHandler).not.toHaveBeenCalled();
      });
    });

    describe('invalid through field validators', () => {
      const createInvalidField = (): IMockField => {
        const field = createMockField('invalidField', 'Invalid field');
        field.state.validate = jest.fn().mockResolvedValue({
          isValidating: false,
          valid: false,
          error: 'foobar',
        });

        return field;
      };

      it('should trigger a submit-invalid event', async () => {
        const { fieldEventResult } = await setupSubmit({ customField: createInvalidField() });
        expect(fieldEventResult.notifyListeners).toHaveBeenCalledWith('_form', 'submit-invalid');
      });

      it('should not call the onSubmit prop', async () => {
        const onSubmitHandler = jest.fn();
        await setupSubmit({
          props: { onSubmit: onSubmitHandler },
          customField: createInvalidField(),
        });
        expect(onSubmitHandler).not.toHaveBeenCalled();
      });
    });

    describe('context busy state', () => {
      const testBusyState = async (result: IMockResult, expected: boolean): Promise<void> => {
        await simulateSubmitEvent(result.current);
        expect(result.current.busy).toBe(expected);
      };

      it('should not be busy if the onSubmit callback returns immediately', async () => {
        const onSubmitHandler = jest.fn();
        const { result } = setup({ props: { onSubmit: onSubmitHandler }});
        await testBusyState(result, false);
      });

      it('should not be busy if there is no onSubmit callback', async () => {
        const { result } = setup();
        await testBusyState(result, false);
      });

      describe('busy prop override set to true', () => {
        it('should be always busy', () => {
          const { result } = setup({ props: { busy: true }});
          expect(result.current.busy).toBeTruthy();
        });

        it('should be busy if the onSubmit callback returns immediately', async () => {
          const onSubmitHandler = jest.fn();
          const { result } = setup({ props: { onSubmit: onSubmitHandler, busy: true }});
          await testBusyState(result, true);
        });

        it('should be busy if there is no onSubmit callback', async () => {
          const { result } = setup({ props: { busy: true }});
          await testBusyState(result, true);
        });
      });

      describe('async onSubmit callback', () => {
        const createSlowOnSubmit = (): () => Promise<void> => {
          return async (): Promise<void> => new Promise<void>(
            (resolve: Function) => setTimeout(
              (): void => { resolve(); },
              1000,
            ),
          );
        };

        beforeAll(() => {
          jest.useFakeTimers();
        });

        afterAll(() => {
          jest.useRealTimers();
        });

        it('should be busy after invoking onSubmit', async () => {
          const { result } = setup({ props: { onSubmit: createSlowOnSubmit() }});
          await simulateSubmitEvent(result.current);

          expect(result.current.busy).toBe(true);
        });

        it('should not be busy after onSubmit finished', async () => {
          const { result } = setup({ props: { onSubmit: createSlowOnSubmit() }});
          await simulateSubmitEvent(result.current);

          await act(async () => {
            jest.runAllTimers();
          });

          expect(result.current.busy).toBe(false);
        });

        it('should be busy after onSubmit finished if the busy prop is set to true', async () => {
          const { result } = setup({ props: { onSubmit: createSlowOnSubmit(), busy: true }});
          await simulateSubmitEvent(result.current);

          await act(async () => {
            jest.runAllTimers();
          });

          expect(result.current.busy).toBe(true);
        });
      });
    });

    describe('reset on successful submit', () => {
      it('should call the onReset prop on submit when resetOnSubmit is true', async () => {
        const onResetHandler = jest.fn();
        const { result } = setup({ props: { onReset: onResetHandler, resetOnSubmit: true }});
        await simulateSubmitEvent(result.current);

        expect(onResetHandler).toHaveBeenCalled();
      });
    });
  });

  describe('onReset handling', () => {
    const simulateResetEvent = (formContext: IFormContext): void => {
      act(() => {
        formContext.reset();
      });
    };

    it('should reset all fields', () => {
      const unitField = createMockField('unitField', 'Unit field');
      const unitGroup = createMockField('unitGroup', 'Unit group', true);
      const unitSubField = createMockField(`${unitGroup.name}.subField`, 'Sub field');
      const mockFields = new Map<string, IFieldState>([
        [ unitField.name, unitField.state ],
        [ unitGroup.name, unitGroup.state ],
        [ unitSubField.name, unitSubField.state ],
      ]);

      const forEachFieldState = jest.fn().mockImplementation((cb) => {
        mockFields.forEach(cb);
      });

      const { result } = setup({ fieldStatesOverride: { forEachFieldState }});

      simulateResetEvent(result.current);
      mockFields.forEach(item => expect(item.reset).toHaveBeenCalled());
    });

    it('should call the onReset prop', () => {
      const onResetHandler = jest.fn();
      const { result } = setup({ props: { onReset: onResetHandler }});
      simulateResetEvent(result.current);

      expect(onResetHandler).toHaveBeenCalled();
    });
  });
});
