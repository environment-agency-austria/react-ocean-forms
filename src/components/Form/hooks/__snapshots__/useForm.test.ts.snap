// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForm invalid field registration case 1 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useForm invalid field registration case 2 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useForm invalid field registration case 3 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useForm invalid field registration case 4 of 5 props 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useForm invalid field registration case empty state 1`] = `"[Form] registerField: invalid field state given"`;

exports[`useForm invalid field registration case invalid field name 1`] = `"[Form] registerField: name is required"`;

exports[`useForm invalid field registration case no parameters 1`] = `"[Form] registerField: name is required"`;

exports[`useForm invalid field registration case no state 1`] = `"[Form] registerField: field state is required"`;
